### PR TITLE
Rdar 30851899 bridging pch mergemodule mismatch swift 4.0 branch

### PR DESCRIPTION
**Pull request by graydon**
**Explanation**: Fix a bridging PCH crash caused by a clang bug

**Scope**: Causes failure-to-compile (appears as a deserialization error) unless user disables bridging PCH. Triggered by legal-but-clumsy misconfiguration in workspace, seemingly as-performed by cocoapods in unknown number of cases. Might be quite widespread.

**Radar**: 30851899

**Risk**: relatively low, it's not even clear to me that a user can detect it in C/ObjC, and it only makes an under-defined corner case of the module system a bit more-deterministic; it's just a corner-case that swift happens to hit when importing definitions from a nest of header symlinks, as produced by cocoapods.

**Testing**: Swift has a pending test for it https://github.com/apple/swift/pull/10658 ; at the clang level, it can't easily be checked in ObjC, but the misbehaviour can trigget an assertion in clang, which there's an accompanying test for here.

### CODE REVIEW

**Comment by graydon** [comment]
@swift-ci please test

File: include/clang/AST/Redeclarable.h [hunk 1]
```diff
@@ -21,6 +21,60 @@
...
+//      |            |       |              |       |              |
+//      +------------+       +--------------+       +--------------+
+//      | data       | defn  | data         |  defn | data         |
+//      |            | ----> |              | <---- |              |
+//      +------------+       +--------------+       +--------------+
+//        |                     |     ^                  ^
+//        |                     |defn |                  |
+//        | link                +-----+                  |
+//        +-->-------------------------------------------+
+
```
Comment (bcardosolopes) [inline include/clang/AST/Redeclarable.h:77]: This is very nice!

**Review by bcardosolopes — APPROVED** [review]
